core/lib/zksync_core/src/utils/mod.rs

```diff
@@ -87,6 +87,33 @@ pub(crate) async fn projected_first_l1_batch(
     Ok(snapshot_recovery.map_or(L1BatchNumber(0), |recovery| recovery.l1_batch_number + 1))
 }
 
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct BlockStartInfo {
+    /// Projected number of the first locally available miniblock. This miniblock is **not**
+    /// guaranteed to be present in the storage!
```

Comment:
Also, stupid question: looks like we have some places where the fact that there may be some state _but_ no miniblocks/batches in the storage after the snapshot sync is an edge case. Would it make sense to include the last L1 batch into the snapshot? This way we'll always be sure that there is at least one miniblock/batch in the storage and we'll be able to use it as an invariant.
The cost of +1 L1 batch sounds negligible in theory, but it can simplify stuff quite a bit.

Reply:
We've discussed this with @tomg10 and @RomanBrodetski before and ultimately decided against it. The reasoning for it is that a miniblock / L1 batch header in Postgres *sort of* assumes that there's information related to it (transactions, events, etc.) is also present in Postgres. This would not be the case for the snapshot miniblock / L1 batch, so if we store their headers in Postgres, it would make more difficult to reason about storage invariants.

Reply:
But why won't we ship events/transactions/etc as well? For a single batch, it shouldn't add too much data.
Basically what I had in mind is that we'll _really_ fully sync a single block in addition to the rest of snapshot contents.

Reply:
I would be *somewhat* against for a couple of reasons:

- It seems we cannot ensure that the data (events, L2-to-L1 logs etc.) produced by transaction execution in the snapshot miniblock or L1 batch is genuine. To ensure it's genuine, we'd need to actually execute transactions, which would require a VM state snapshot for the previous L1 batch 🙃 As long as we keep snapshot to *just* the VM state (i.e., 3 pieces of info: storage logs, initial writes and factory deps), it's moderately easy to authenticate it using root hash published on L1. AFAIU, the same mechanism could be used to authenticate factory dependencies (since their hashes are a part of the VM state). Authenticity of initial writes seem to be most complicated, but still doable. Or is it possible to verify authenticity of all data associated with a miniblock / L1 batch?
- More banal argument: If we keep snapshot data to minimum, it's easier to maintain. In your proposal, we'd need to ship *all* data associated with a miniblock / L1 batch, which can evolve with time, so old snapshots might cease to work, non-updated nodes might not understand new snapshots etc.

Reply:
Fair enough.